Add support for membership, exposing a userlist and joins/parts callbacks

smt923 opened the pull request:
- added support for twitch irc [membership](https://dev.twitch.tv/docs/irc/membership/), keeping a userlist that's updated on every `NAMES`, `JOIN` and `PART`
- exposed a new `Client` method which returns the current userlist for a given channel:
```go
client.Userlist(channel string) []string
```
- added `OnUserJoin` and `OnUserPart` callbacks, giving the channel and user to the caller, allowing for things like join/part messages or other custom handling _(the user param here has to be a string here as the join/part messages contain next to no info, so I couldn't use the `User` struct unless we want to fill it with mostly dummy/assumed info)_

also added tests for everything and a new test helper for string slices, all tests pass for me and everything in some trivial tests, should close #20

gempir commented:
Thank you very much for your Contribution. I will review this on the weekend.

gempir commented on client.go:
```diff
@@ -208,6 +224,12 @@ func (c *Client) Connect() error {
 }
 
+// Userlist returns the userlist for a given channel
+func (c *Client) Userlist(channel string) []string {
```
This method should also return an error IMO that signals when a userlist couldn't be found for a channel, since we haven't joined that channel. An empty list would be confusing.

gempir commented on client.go:
```diff
@@ -147,6 +162,7 @@ func (c *Client) Depart(channel string) {
 
 	c.channelsMtx.Lock()
 	delete(c.channels, channel)
+	delete(c.channelUserlist, "channel")
```
why is this a string of channel?

gempir commented on client.go:
```diff
@@ -291,6 +314,55 @@ func (c *Client) handleLine(line string) {
+	}
+}
+
+func removeElement(elem string, slice []string) []string {
```
This is very inefficent and should be just another map aswell. So a map inside a map in total. The value doesn't really matter in case of the second map.

gempir commented on client.go:
```diff
@@ -291,6 +314,55 @@ func (c *Client) handleLine(line string) {
+	return slice[:j]
+}
+
+func isInSlice(elem string, slice []string) bool {
```
Same as above, this is not efficent. A map would be better suited.

smt923 commented:
Thanks a lot for the feedback, maps are a lot easier and I'm not sure why I didn't think to use them first, also fixed some of my other little mistakes and updated the tests, all seems to work as well as before

gempir commented on client.go:
```diff
@@ -208,6 +224,14 @@ func (c *Client) Connect() error {
+// Userlist returns the userlist for a given channel
+func (c *Client) Userlist(channel string) (map[string]bool, error) {
+	if value, ok := c.channelUserlist[channel]; ok {
+		return value, nil
```
This is a lot better now, but I would prefer if the function converts our internal map into a slice because that's a lot more usable. For Performance reasons we would keep storing it as a map

smt923 commented:
converting it to a slice definitely makes more sense, updated the tests & documentation as well

gempir commented on README.MD:
```diff
@@ -60,6 +60,7 @@ These are the available methods of the client so you can get your bot going:
 	func (c *Client) Whisper(username, text string)
 	func (c *Client) Join(channel string)
 	func (c *Client) Depart(channel string)
+	func (c *Client) Userlist(channel string) ([]string, err)
```
this should be error, err is not a type

gempir commented:
Tiny thing i just noticed in the readme, then I can merge.

smt923 commented:
ah oops, thought you wanted me to grab it, hopefully it's okay
